airbyte-cdk/python/airbyte_cdk/v2/state/state_manager.py

```diff
@@ -0,0 +1,107 @@
+from abc import abstractmethod, ABC
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import List, Tuple, Optional, Mapping, Any, Generic, TypeVar, Iterable
+
+from airbyte_protocol.models import ConfiguredAirbyteCatalog
+from pydantic import BaseModel, Field
+
+from .datetime_range_tracker import DatetimeRangeTracker
+from ..concurrency.partition_descriptors import PartitionDescriptor, PartitionGenerator
+from ..state_obj import State
+
+StateType = TypeVar('StateType', bound=State)
+PartitionType = TypeVar('PartitionType', bound=PartitionDescriptor)
+
+
+class StateManager(ABC, Generic[StateType, PartitionType]):
+    def observe(self, record: Mapping[str, Any]) -> Optional[StateType]:
+        """ If the state was updated as a result of observing this record, returns a state object"""
+        pass
+
+    @abstractmethod
+    def notify_partition_complete(self, partition_descriptor: PartitionType) -> StateType:
+        """ Called when a partition has successfully synced"""
+
+    @abstractmethod
+    def get_state(self) -> StateType:
+        """Returns the current checkpointable state. """
+
+    @staticmethod
+    @abstractmethod
+    def from_state(state: StateType = None):
+        """Return a new instance of the state manager initialized with the state object"""
+
+class LegacyStateManager(StateManager[StateType, PartitionType]):
+
+    def notify_partition_complete(self, partition_descriptor: PartitionType) -> StateType:
+        return EmptyState()
+
+    def get_state(self) -> StateType:
+        return EmptyState()
+
+    @staticmethod
+    def from_state(state: StateType = None):
+        pass
+
+
+class DatetimeState(State, BaseModel):
+    copied_ranges: List[Tuple[datetime, datetime]] = Field(...)
+
+    @staticmethod
+    def from_dict(d: Mapping[str, Any]) -> StateType:
+        return DatetimeState.parse_obj(d)
+
+    def to_dict(self) -> Mapping[str, Any]:
+        return self.dict(exclude_unset=True)
+
+class EmptyState(State):
+    @staticmethod
+    def from_dict(d: Mapping[str, Any]) -> StateType:
+        return EmptyState()
+
+    def to_dict(self) -> Mapping[str, Any]:
+        return {}
+
+
+@dataclass
+class DatetimePartitionDescriptor(PartitionDescriptor):
+    start_datetime: datetime
+    end_datetime: datetime
+
+
+class DatetimePartitionGenerator(PartitionGenerator[DatetimePartitionDescriptor, DatetimeState]):
```

Comment:
I'm not using this right now. It's a nice abstraction, but we can't go from stream_slices -> DatetimePartitionDescriptor. Using this would require updating the connector